packages/design-system/src/components/primitives/dnd/use-drop.ts

```diff
@@ -0,0 +1,293 @@
+        ] as Rect | undefined;
+
+        const placement =
+          toGlobalCoordinates(
+            getPlacementBetween(closestChildRect, neighbourRect)
+          ) ||
+          toGlobalCoordinates(
+            getPlacementNextTo(
+              parentRect,
+              closestChildRect,
+              childrenOrientation,
+              indexAdjustment > 0 ? "forward" : "backward"
+            )
+          ) ||
+          getPlacementInside(parentRect, childrenOrientation);
+
+        const dropTarget: DropTarget<Data> = {
+          ...partialDropTarget,
+          rect: parentRect,
+          indexWithinChildren,
+          placement,
+        };
+
+        state.current.dropTarget = dropTarget;
+        latestProps.current.onDropTargetChange(dropTarget);
+      }
+    };
+
+    const detectTarget = () => {
+      const {
+        edgeDistanceThreshold = 3,
+        isDropTarget,
+        swapDropTarget,
+      } = latestProps.current;
+
+      const { pointerCoordinates } = state.current;
+      const root = rootRef.current;
+
+      // @todo: Cache this?
+      // Not expensive by itself, but it may call isDropTarget multiple times.
+      let candidate =
+        root &&
+        findClosestDropTarget({
+          root,
+          initialElement:
+            pointerCoordinates &&
+            toUseableElement(
+              document.elementFromPoint(
+                pointerCoordinates.x,
+                pointerCoordinates.y
+              )
+            ),
+          isDropTarget,
+        });
+
+      // To avoid calling swapDropTarget unnecessarily on every pointermove
+      const isNewCandidate =
+        candidate?.element !== state.current.lastCandidateElement;
+      state.current.lastCandidateElement = candidate?.element;
+      const candidateIsNearEdge =
+        candidate &&
+        pointerCoordinates &&
+        isNearEdge(
+          pointerCoordinates,
+          edgeDistanceThreshold,
+          candidate.element.getBoundingClientRect()
+        );
+      const isNewIsNearEdge =
+        candidateIsNearEdge !== state.current.lastCandidateIsNearEdge;
+      state.current.lastCandidateIsNearEdge = candidateIsNearEdge;
+      if (!isNewCandidate && !isNewIsNearEdge && state.current.dropTarget) {
+        // Still need to call setDropTarget to update rect and/or placement.
+        // Because indexWithinChildren might have changed,
+        // or parent coordinates might have moved in case of a scroll
+        setDropTarget(state.current.dropTarget);
+        return;
+      }
+
+      // eslint-disable-next-line no-constant-condition
+      while (true) {
+        const swappedTo = swapDropTarget(
+          candidate &&
+            pointerCoordinates && {
+              ...candidate,
+              nearEdge: isNearEdge(
+                pointerCoordinates,
+                edgeDistanceThreshold,
+                candidate.element.getBoundingClientRect()
```

Comment:
this will be potentially an expensive call since its going to trigger reflow, how often do you think this is going to be called in a large tree?

Reply:
You tested it with a loop in a console back when we paired. I'll keep an eye on this when I'll do the test with 1000+ instances.

Reply:
I think I was testing get element from point performance